src/nypl/mod.rs

```diff
@@ -0,0 +1,204 @@
+        if data.uri.starts_with(NYPL_IMAGE_VIEW_PREFIX) {
+            self.assert(data.uri.contains(NYPL_IMAGE_VIEW_PREFIX))?;
+            let image_id = data.uri.replace(NYPL_IMAGE_VIEW_PREFIX, "");
+            let meta_uri = format!(nypl_meta_format_tpl!(), image_id);
+            Err(DezoomerError::NeedsData { uri: meta_uri })
+        } else {
+            self.assert(data.uri.contains(NYPL_META_PREFIX))?;
+            let DezoomerInputWithContents { uri, contents } = data.with_contents()?;
+            let iter = iter_levels(uri, contents).map_err(DezoomerError::wrap)?;
+            Ok(iter.into_zoom_levels())
+        }
+    }
+}
+
+fn arcs<T>(v: T) -> impl Iterator<Item=Arc<T>> {
+    successors(Some(Arc::new(v)), |x| Some(Arc::clone(x)))
+}
+
+fn iter_levels(uri: &str, contents: &[u8])
+               -> Result<impl Iterator<Item=Level> + 'static, NYPLError> {
+    let base = _get_image_url_from_meta_url(uri);
+    let meta = Metadata::try_from(contents)?;
+    let levels =
+        (0..meta.levels).zip(arcs(base)).zip(arcs(meta))
+            .map(|((level, base), metadata)|
+                Level { metadata, base, level });
+    Ok(levels)
+}
+
+#[derive(PartialEq)]
+struct Level {
+    metadata: Arc<Metadata>,
+    base: Arc<String>,
+    level: u32,
+}
+
+impl Debug for Level {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "NYPLImage")
+    }
+}
+
+impl TilesRect for Level {
+    fn size(&self) -> Vec2d {
+        let reverse_level = self.metadata.levels - self.level - 1;
+        self.metadata.size / 2_u32.pow(reverse_level)
+    }
+
+    fn tile_size(&self) -> Vec2d { self.metadata.tile_size }
+
+    fn tile_url(&self, Vec2d { x, y }: Vec2d) -> String {
+        let Vec2d { x: _width, .. } = self.size().ceil_div(self.tile_size());
+        format!("https://access.nypl.org/image.php/{id}/tiles/0/12/{x}_{y}.png",
+                id = self.base,
+                x = x,
+                y = y,
+        )
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Metadata {
+    size: Vec2d,
+    tile_size: Vec2d,
+    levels: u32,
+}
+
+impl FromStr for Metadata {
+    type Err = NYPLError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use NYPLError::*;
+        let _parsed = json::parse(s);
+        if _parsed.is_err(){
+            return Err(JsonError);
+        }
+        let parsed = _parsed.unwrap();
+        let meta = parsed["configs"]["0"].to_owned();
+        let width = meta["size"]["width"].as_str().unwrap()
+            .parse::<u32>().unwrap();
+        let height = meta["size"]["height"].as_str().unwrap()
+            .parse::<u32>().unwrap();
+        let _tile_width = meta["tilesize"].as_str().unwrap()
+            .parse::<u32>().unwrap();
+        let size = Vec2d { x: width, y: height };
+        let tile_size = Vec2d{x: _tile_width, y: _tile_width};
+        let levels= 1;
+        Ok(Metadata { size, tile_size, levels })
```

Comment:
We can remove that once we use serde_json

Reply:
I use the same logic as the old one.
I prefer to not build a strut to parse it (because it will not be used at other place).